### PR TITLE
Allow overriding the received (this) of a jsg::Function, update forEach impls

### PR DESCRIPTION
**Pull request by jasnell**
The jsg::Function will automatically set the receiver when mapping
from a JS function to jsg::Function. This is not always what we want.
In many cases we have to be able to override the receiver on a case
by case basis. This pr introduces the ability to change the receiver
on a jsg::Function, and updates the various forEach method impls to
use jsg::Function instead of directly using v8::Function calls.

### CODE REVIEW

File: src/workerd/jsg/function.h
Comment (ohodson) [inline src/workerd/jsg/function.h:239]: This is my ignorance, but is it legitimate for instances of non-Object types to feature as a receiver in JS?

**Comment by ohodson** [comment]
lgtm modulo my naive question about value vs object for the receiver. Probably good to have a second opinion though.

File: src/workerd/jsg/function.h
Comment (jasnell) [reply src/workerd/jsg/function.h:239]: Yep. When a primitive is passed, it will be mapped to the the boxed wrapper.

```
function foo() { console.log(this); }
foo.call(123); // prints `[Number 123]`, this will be a NumberObject
foo.call('hello'); // prints `[String "hello"]`, this will be a StringObject
```

**Review by ohodson — APPROVED** [review]

File: src/workerd/api/url.h
Comment (Warfields) [inline src/workerd/api/url.h:166]: Possibly just learned something about c++, just want to verify before I make any assumptions. Can you always just list types, and then let the implementation name the variables or is this a special case?

**Review by Warfields — APPROVED** [review]

File: src/workerd/api/url.h
Comment (jasnell) [reply src/workerd/api/url.h:166]: Yep. I know there are differences of opinion on this about whether it is desirable but personally I prefer the more abbreviated form in declarations when the intent is clear.

**Pull request merged**
